ui/src/table.tsx

```diff
@@ -241,11 +241,10 @@ const
     'on-hover': Fluent.CheckboxVisibility.onHover,
     'hidden': Fluent.CheckboxVisibility.hidden,
   },
-  groupByF = function <T extends Dict<any>>(arr: T[], key: S): Dict<any> {
-    return arr.reduce((rv, x: T) => {
-      (rv[x[key]] = rv[x[key]] || []).push(x)
-      return rv
-    }, {} as Dict<any>)
+  groupByF = function <T extends Dict<any>>(items: T[], filteredItems: T[], key: S): Dict<any> {
+    const groupedItems = [...new Set(items.map(item => item[key]))].reduce((arr, val) => { arr[val] = []; return arr }, [] as Dict<any>)
```

Comment:
Previously I had it implemented like this:
```tsx
    const groupedItems = items.reduce((arr, item) => {
      if (!arr[item.status]) arr[item.status] = []
      return arr
    }, [] as any)
    filteredItems.forEach(item => groupedItems[item[key]].push(item))
    return groupedItems
```

Also there is another alternative using `forEach` only:
```tsx
    const groupedItems: Dict<any> = []
    items.forEach(item => { if (!groupedItems[item[key]]) groupedItems[item[key]] = [] })
    filteredItems.forEach(item => groupedItems[item[key]].push(item))
    return groupedItems
 ```

Reply:
While a little faster, still not as fast as it was before as both options iterate over all items, which seems unnecessary.

Reply:
Let's say you have following `items` and `filteredItems`:
```tsx
const items=[
  {text: 'Aaaa', status: 'Open'},
  {text: 'Bbbb', status: 'Open'},
  {text: 'Cccc', status: 'Closed'},
  {text: 'Dddd', status: 'Open'},
  {text: 'Eeee', status: 'Closed'},
]

const filteredItems=[
  {text: 'Aaaa', status: 'Open'},
  {text: 'Bbbb', status: 'Open'},
  {text: 'Dddd', status: 'Open'},
]
```

Now what you want is to group items by `text`.
The expected output is to get a dictionary of 5 groups:
```tsx
const groups = [
'Aaaa': [{text: 'Aaaa', status: 'Open'}], 
'Bbbb': [{text: 'Bbbb', status: 'Open'}], 
'Cccc': [], 
'Dddd': [{text: 'Dddd', status: 'Open'}], 
'Eeee': []
]
```

How are you supposed to get this output without iterating through all items?

___ 

I've improved my previous solution by iterating through `items` only (no need to iterate through `filteredItems` separately).
```tsx
  groupByF = function <T extends Dict<any>>(items: T[], filteredItems: T[], key: S): Dict<any> {
    return items.reduce((arr, item, idx) => {
      if (!arr[item[key]]) arr[item[key]] = []
      if (idx < filteredItems.length) {
        const groupName = filteredItems[idx][key]
        if (arr[groupName]) arr[groupName].push(filteredItems[idx])
        else arr[groupName] = [filteredItems[idx]]
      }
      return arr
    }, [] as Dict<any>)
  },
```

Reply:
> How are you supposed to get this output without iterating through all items?

What's the point of iterating through all the items every single time we need to recalculate the groups when we know the result will be the same as long as `items` did not change?

Reply:
> as long as `items` did not change

The result won't be the same also if `key` has changed. 

But I understand your point that we don't need to iterate through all items every time we need to recalculate groups in case `key` or `items` are not changed.

In that case I suggest holding a separate list of all groups (including empty ones) which will be recalculated every time `items`, `groupByKey` or `m.groups` has changed. Wdyt?

Reply:
> I suggest holding a separate list of all groups (including empty ones) which will be recalculated every time items, groupByKey or m.groups has changed

Yes if "all groups" means the groups related to the currently selected column, not all possible groups across all columns.

Reply:
> means the groups related to the currently selected column

Yep, this is what I mean.